### PR TITLE
Fix cloning per subdirectory wrt #5411

### LINKED ISSUE #5411: extra-deps' clones as many times the repository as there are subdirs
### General summary/comments (optional)
When I specify, in extra-deps, a repository with many subdirs, the repository is clone many times.

### Steps to reproduce

Build a project with this `package.yml` part:

```
extra-deps:
- git: https://github.com/theam/aws-lambda-haskell-runtime.git
  commit: 6aef017c1621c828a47dc98eae12d84818a8d45c
- git: https://github.com/brendanhay/amazonka.git
  commit: 020bc7bde47bb235e448c76088dc44d6cec13e9b
  subdirs:
    - gen
    - core
    - test
    - examples
    - amazonka
```

### Expected

I expect only one cloning.

### Actual

The repository is cloned once by subdir.

[Partial output](https://gist.github.com/blackheaven/23af3091da8c7ad4b73eb07416edcb7f)

### Stack version

```
$ stack --version
Version 2.3.3, Git revision cb44d51bed48b723a5deb08c3348c0b3ccfc437e x86_64 hpack-0.33.0
```

### Method of installation

```
curl -sSL https://get.haskellstack.org/ | sh
```

### PR DESCRIPTION
**Pull request by hasufell**
Fixes #5411

Depends on https://github.com/commercialhaskell/pantry/pull/41

----

on cardano-wallet:
```
$ /home/hasufell/git/stack/dist-newstyle/build/x86_64-linux/ghc-8.10.7/stack-2.7.4/build/stack/stack build
Cloning 46ce81050f4015abd40c6bdf0a1ed30fa54e375a from https://github.com/input-output-hk/cardano-addresses
Cloning 7497a29cb998721a9068d5725d49461f2bba0e7a from https://github.com/input-output-hk/optparse-applicative
Cloning cb0f19c85e5bb5299839ad4ed66af6fa61322cc4 from https://github.com/input-output-hk/cardano-base
Cloning 07397f0e50da97eaa0575d93bee7ac4b2b2576ec from https://github.com/input-output-hk/cardano-crypto
Cloning 6aa1cd0a64a464371b94d4ac182e7e2cddc83a36 from https://github.com/input-output-hk/cardano-ledger-specs
Cloning 4c59442958072657812c6c0bb8e0b4ab85ce1ba2 from https://github.com/input-output-hk/cardano-node
Cloning 12925934c533b3a6e009b61ede555f8f26bac037 from https://github.com/input-output-hk/cardano-sl-x509
Cloning ee59880f47ab835dbd73bea0847dab7869fc20d8 from https://github.com/input-output-hk/flat
Cloning cde90a2b27f79187ca8310b6549331e59595e7ba from https://github.com/input-output-hk/goblins
Cloning edf6945007177a638fbeb8802397f3a6f4e47c14 from https://github.com/input-output-hk/hedgehog-extras
Cloning 808724ff8a19a33d0ed06f9ef59fbd900b08553c from https://github.com/input-output-hk/iohk-monitoring-framework
Cloning 6a92d7853ea514be8b70bab5e72077bf5a510596 from https://github.com/shmish111/purescript-bridge.git
Cloning 877ce057ff6fb086474c8eaad53f2b7f0e0fce6b from https://github.com/input-output-hk/ouroboros-network
Cloning edc6d4672c41de4485444122ff843bc86ff421a0 from https://github.com/input-output-hk/plutus
Cloning a76104490499aa72d40c2790d10e9383e0dbde63 from https://github.com/shmish111/servant-purescript.git
Cloning 3825d3abf75f83f406c1f7161883c438dac7277d from https://github.com/input-output-hk/Win32-network
Preparing to install GHC (tinfo6) to an isolated location.
This will not interfere with any system-level installation.
Downloaded ghc-tinfo6-8.10.5.
Installed GHC.
WARNING: Ignoring text's bounds on ghc-prim (>=0.2 && <0.6); using ghc-prim-0.6.1.
Reason: allow-newer enabled.
WARNING: Ignoring text's bounds on template-haskell (>=2.5 && <2.16); using template-haskell-2.16.0.0.
Reason: allow-newer enabled.
WARNING: Ignoring recursion-schemes's bounds on th-abstraction (>=0.2.4 && <0.4); using th-abstraction-0.4.2.0.
Reason: allow-newer enabled.
WARNING: Ignoring size-based's bounds on template-haskell (>=2.5 && <2.16); using template-haskell-2.16.0.0.
Reason: allow-newer enabled.
WARNING: Ignoring lens's bounds on Cabal (>=1.10 && <3.3); using Cabal-3.4.0.0.
Reason: allow-newer enabled.
WARNING: Ignoring ral's bounds on fin (^>=0.1.1); using fin-0.2.
Reason: allow-newer enabled.
WARNING: Ignoring Unique's bounds on extra (>=1.6.2 && <1.7); using extra-1.7.9.
Reason: allow-newer enabled.
WARNING: Ignoring Unique's bounds on hashable (>=1.2.6 && <1.3); using hashable-1.3.0.0.
Reason: allow-newer enabled.
WARNING: Ignoring ouroboros-consensus-byron's bounds on formatting (>=6.3 && <6.4); using formatting-7.1.3.
Reason: allow-newer enabled.
WARNING: Ignoring async-timer's bounds on unliftio-core (>=0.1.1.0 && <0.2); using unliftio-core-0.2.0.1.
Reason: allow-newer enabled.
[1 of 2] Compiling Main             ( /home/hasufell/.stack/setup-exe-src/setup-mPHDZzAJ.hs, /home/hasufell/.stack/setup-exe-src/setup-mPHDZzAJ.o )
[2 of 2] Compiling StackSetupShim   ( /home/hasufell/.stack/setup-exe-src/setup-shim-mPHDZzAJ.hs, /home/hasufell/.stack/setup-exe-src/setup-shim-mPHDZzAJ.o )
Linking /home/hasufell/.stack/setup-exe-cache/x86_64-linux-tinfo6/tmp-Cabal-simple_mPHDZzAJ_3.2.1.0_ghc-8.10.5 ...
StateVar                         > configure
...
```

### CODE REVIEW

File: stack.yaml [hunk 1]
```diff
@@ -5,6 +5,10 @@ packages:
 
 extra-deps:
 - rio-0.1.21.0@rev:0
+# TODO: temporary
```
Comment (snoyberg) [inline stack.yaml:8]: Please update this, as well as the other stack config files

Comment (hasufell) [reply stack.yaml]: Done

**Review by snoyberg — APPROVED** [review]
LGTM, thanks!

**Comment by mbj** [comment]
Thanks all for this work! Greatly apprechiated on my CI times.

**Pull request merged**